File: optimal/4.R
# Largest palindrome product

Args   <- commandArgs()
digits <- ifelse(is.na(Args[6]), 3, as.numeric(Args[6]))

high_num <- 10 ^ digits - 1
low_num  <- 10 ^ (digits - 1)

is_palindrome <- function(number) {
    digit_chars <- strsplit(as.character(as.integer(number)), split=NULL)[[1]]
    reversed_number <- as.integer(paste(rev(digit_chars), collapse=''))
    return(number == reversed_number)
}

max <- 0

num1 <- high_num
while ( num1 >= low_num ) {
    num2 <- num1
    decrease <- 1
    if ( num2 %% 11 ) {
        num2 <- num2 %/% 11 * 11
        decrease <- 11
    }
    while ( num2 >= low_num ) {
        product <- num1 * num2
        if ( product > max & is_palindrome(product) ) {
            max <- product
        }
        num2 <- num2 - decrease
    }
    num1 <- num1 - 1
}

cat(max, fill=TRUE)

File: optimal/4.js
#!/usr/bin/env node

// Largest palindrome product

var program = require('commander');

program
    .version('0.1.0')
    .description('Largest palindrome product')
    .option('-d, --digits <int>',
        'The number of digits in the numbers to be multiplied', Number, 3)
    .parse(process.argv);

var high_num = '9'.repeat(program.digits);
var low_num = Math.pow(10, program.digits - 1);
var max = 0;
for (var num1 = high_num; num1 >= low_num; num1--) {
    var num2 = num1;
    var decrease = 1;
    if (num1 % 11) {
        num2 = Math.floor(num2 / 11) * 11;
        decrease = 11;
    }
    while (num2 >= low_num) {
        var product = num1 * num2;
        if (product > max && is_palindrome(product)) {
            max = product;
        }
        num2 -= decrease;
    }
}

console.log(max);

function is_palindrome(number) {
    return number.toString().split('').reverse('').join('') == number;
}
